sw/otbn/crypto/sha512_compact.s

```diff
@@ -0,0 +1,435 @@
+ * variant at the cost of additional computation time.
+ *
+ * Terminology within the comments in this library is based (as much as
+ * possible) on the terminology of FIPS 180-4.
+ *
+ * The functions named with the greek sigma have been renamed:
+ *   - sigma_lowercase(x) -> s(x)
+ *   - sigma_uppercase(x) -> S(x)
+ *
+ * Upercase W_i denotes the i-th word from the message schedule.
+ * Uppercase M_i denotes the i_th message word of the current chunk.
+ * Uppercase K_i denotes the i-th round constant.
+ *
+ * Note on SHA-384: SHA-384 uses the same base algorithm as SHA-512. The
+ * only differences are:
+ *    - different intial hash values for each variant
+ *    - for SHA-384 the message digest is truncated to 384 bit
+ *      (only the first 6 final hash values are used)
+ * Both aspects are out of scope of this implementation. Therefore, this
+ * implementation provides the base algorithm for bose variants.
+ */
+
+
+.section .text
+
+
+/**
+ * Compute SHA-512 (or SHA-384) hash
+ *
+ * Updates the SHA-512 state for n subsequent 1024-bit chunks of a
+ * pre-formatted message.
+ *
+ * The message is expected in dmem in a pre-processed format:
+ *  - The message has been padded according to the SHA-512 standard.
+ *  - The padded message has been broken down into 64-bit sized big-endian
+ *    words. I.e. for a message stored at dmem address a, the expected
+ *    formating for the first 16 message bytes is as follows
+ *    (where mn denotes the n-th message byte):
+ *    |a+15|a+14|a+13|a+12|a+11|a+10|a+9|a+8|a+7|a+6|a+5|a+4|a+3|a+2|a+1|a+0|
+ *    |  m8|  m9| m10| m11| m12| m13|m14|m15| m0| m1| m2| m3| m4| m5| m6| m7|
+ *
+ * The state variables H[0] to H[7] are expected in dmem in 8 subsequent
+ * 256-bit memory cells, where each state variable occupies the lower 64 bit
+ * of such a cell. For the state stored at dmem address a the expected format
+ * is as follows:
+ *  dmem[a +    ][63:0]: H[0]
+ *  dmem[a +  32][63:0]: H[1]
+ *  dmem[a +  64][63:0]: H[2]
+ *  dmem[a +  96][63:0]: H[3]
+ *  dmem[a + 128][63:0]: H[4]
+ *  dmem[a + 160][63:0]: H[5]
+ *  dmem[a + 192][63:0]: H[6]
+ *  dmem[a + 224][63:0]: H[7]
+ * The upper 192 bits of each cell are clobbered during the execution of the
+ * algorithm but their contents are irrelevant.
+ *
+ * This routine runs in constant time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  dmem[n_chunks]: number of chunks to process in a single go
+ * @param[in]  dmem[dptr_state]: dmem location with state      ][63:0]: H[0]
+ * @param[in]  dmem[dptr_msg]: Pointer to memory location containing the pre-
+ *                               formatted message chunks.
+ *
+ * clobbered registers: w0 to w7, w10, w11, w15 to w27, w30, w31
+ *                      x1, x2, x10, x11 to x17, x20
+ * clobbered flag groups: FG0
+ */
+.globl sha512_compact
+sha512_compact:
+
+  /* w31 = 0; w30 = 1111...1111 */
+  bn.xor  w31, w31, w31
+  bn.subi w30, w31, 1
```

Comment:
Out of curiosity, why do you need `w30 = 1111...1111` here? I only see it used below in rotate instructions, where all its bits are eventually masked out.

Reply:
Good point. The routine for a single round here is the same as in the other version (unrolled). But I just checked, also there it doesn't matter how we set the bits in the unused part of the wregs. and also the all-one vector is not needed anywhere else. So we can just use `w31` in the shifting operations.
It was probably needed in a previous version.
As a poor man's countermeasure we could also use random data but since we don't have any claims on the SCA resistance of these algorithms (other than constant time) I'd suggest let's not do that.
I'll update this PR for bot SHA512 variants to remove the instruction.